### PR TITLE
[SPARK-7299][SQL] Set precision and scale for Decimal according to JDBC metadata instead of returned BigDecimal

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-7299

When connecting with oracle db through jdbc, the precision and scale of `BigDecimal` object returned by `ResultSet.getBigDecimal` is not correctly matched to the table schema reported by `ResultSetMetaData.getPrecision` and `ResultSetMetaData.getScale`.

So in case you insert a value like `19999` into a column with `NUMBER(12, 2)` type, you get through a `BigDecimal` object with scale as 0. But the dataframe schema has correct type as `DecimalType(12, 2)`. Thus, after you save the dataframe into parquet file and then retrieve it, you will get wrong result `199.99`.

Because it is reported to be problematic on jdbc connection with oracle db. It might be difficult to add test case for it. But according to the user's test on JIRA, it solves this problem.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala [hunk 1]
```diff
@@ -363,7 +363,9 @@ private[sql] class JDBCRDD(
             case BooleanConversion    => mutableRow.setBoolean(i, rs.getBoolean(pos))
             case DateConversion       =>
               mutableRow.update(i, DateUtils.fromJavaDate(rs.getDate(pos)))
-            case DecimalConversion    => mutableRow.update(i, rs.getBigDecimal(pos))
+            case DecimalConversion(d) if d != None =>
+              mutableRow.update(i, Decimal(rs.getBigDecimal(pos), d.map(_._1).get, d.map(_._2).get))
+            case DecimalConversion(n) => mutableRow.update(i, Decimal(rs.getBigDecimal(pos)))
```
Comment (squito) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: I think this is a bit clearer as:
```
case DecimalConversion(Some(precision,scale)) =>
  mutableRow.update(i, Decimal(rs.getBigDecimal(pos), precision, scale))
case DecimalConversion(None) => mutableRow.update(i, Decimal(rs.getBigDecimal(pos)))
```

**Comment by squito** [comment]
its a bummer that this is an Oracle specific thing.  Do you think it would be worthwhile to add a test for the behavior in any case, even if in our testing environment it works without the change?  And I'd add also add some comments to the code & test that very clearly indicate this is needed specifically for oracle.

**Comment by viirya** [comment]
Such a test might be confusing? See if others have some comments for that. I would add some comments to the code.

**Comment by squito** [comment]
yeah, I could go either way on the test -- I guess the point was mostly to make sure it still worked with DBs, and also to have some example code somebody could try if they did want to mess with oracle integration.  it was just a thought, I don't feel strongly about it.

btw, thanks for adding the comment.  I had meant something much more brief, just like, "this is needed for oracle, see SPARK-7299" but what you have is good :)

**Comment by rxin** [comment]
@viirya can you bring this up to date? I'd like to merge this.

**Comment by viirya** [comment]
@rxin ok. updated.

**Comment by viirya** [comment]
retest this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by viirya** [comment]
@rxin The tests are passed. Please take a look.

**Comment by rxin** [comment]
Thanks - merged in.

**Pull request closed without merging**